sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -77,4 +79,27 @@ private[sql] object StatFunctions {
       })
     counts.cov
   }
+
+  /** Generate a table of frequencies for the elements of two columns. */
+  private[sql] def crossTabulate(df: DataFrame, col1: String, col2: String): DataFrame = {
+    val tableName = s"${col1}_$col2"
+    val counts = df.groupBy(col1, col2).agg(col(col1), col(col2), count("*")).collect()
```

Comment:
It doesn't matter. you can set a max number (maybe 1 million). If the dataset has less than that, it will just return the entire dataset (at a slightly higher cost to run multiple jobs).